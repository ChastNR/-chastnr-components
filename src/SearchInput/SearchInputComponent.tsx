import { IInputProps, Input } from '../Input';
import { OptionsContainer } from '../Option';
import { ISearchOption } from '../types';

interface ISearchInputProps extends IInputProps {
  isOptionsAvailable?: boolean;
  options?: ISearchOption[];
  renderOption?: (option: ISearchOption, index: number) => React.ReactNode;
}

export const SearchInputComponent: React.FC<ISearchInputProps> = ({
  content,
  isOptionsAvailable,
  options = [],
  renderOption,
  ...restProps
}) => {
  const showOptions = (isOptionsAvailable || options.length > 0) && !!renderOption;

  return (
    <Input
      {...restProps}
      content={
        <>
          {content}
          {showOptions && <OptionsContainer>{options.map(renderOption)}</OptionsContainer>}
        </>
      }
    />
  );
};
