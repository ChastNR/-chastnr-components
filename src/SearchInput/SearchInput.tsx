/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import { Component } from "react";

import { v4 as uuid } from "uuid";

import { InputProps } from "../Input";
import { OptionRenderComponent, SearchOption } from "../types";
import { debounce } from "../utils";
import SearchInputComponent from "./SearchInputComponent";

interface SearchInputProps extends InputProps {
  searchThreshold?: number;
  isOptionsAvailable?: boolean;
  timeout?: number;
  loadOptions: (value: string) => Promise<SearchOption[]>;
  renderOption?: OptionRenderComponent;
}

interface SearchInputState {
  options: SearchOption[];
}

class SearchInput extends Component<SearchInputProps, SearchInputState> {
  private readonly searchInputId: string = uuid();

  private inputEl: HTMLElement | null = null;

  loadOptions = debounce((value: string) => {
    this.props.loadOptions(value).then((options) => {
      this.setState({
        options,
      });
    });
  }, this.props.timeout || 500);

  constructor(props: Readonly<SearchInputProps>) {
    super(props);

    this.state = {
      options: [],
    };

    this.handleChange = this.handleChange.bind(this);
    this.optionHandleClick = this.optionHandleClick.bind(this);
    this.renderOption = this.renderOption.bind(this);
    this.loadOptions = this.loadOptions.bind(this);
  }

  handleChange(value: string): void {
    const { searchThreshold = 3, onChange } = this.props;

    onChange(value);

    this.setState({
      options: [],
    });

    if (value.length < searchThreshold) {
      return;
    }

    this.loadOptions(value);
  }

  optionHandleClick(value: string): void {
    const { onChange } = this.props;

    if (!this.inputEl) {
      this.inputEl = document.getElementById(this.searchInputId);
    }

    this.inputEl?.blur();

    onChange(value);

    this.setState({
      options: [],
    });
  }

  renderOption({ label, value }: SearchOption): JSX.Element {
    const { renderOption: RenderComponent } = this.props;

    const handleClick = () => this.optionHandleClick(`${value}`);

    if (RenderComponent) {
      return (
        <RenderComponent
          className="search_option"
          key={value}
          label={label}
          onClick={handleClick}
        />
      );
    }

    return (
      <div className="search_option" key={value} onClick={handleClick}>
        <div>{label}</div>
      </div>
    );
  }

  render(): JSX.Element {
    const { options } = this.state;

    return (
      <SearchInputComponent
        {...this.props}
        id={this.searchInputId}
        onChange={this.handleChange}
        options={options}
        renderOption={this.renderOption}
      />
    );
  }
}

export default SearchInput;
