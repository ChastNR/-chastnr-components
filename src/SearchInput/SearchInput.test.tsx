import { fireEvent, render } from '@testing-library/react';
import { SearchInput } from './SearchInput';

describe('SearchInput', () => {
  const mockLoadOptions = jest.fn().mockResolvedValue([{ label: 'test', value: 'test' }]);
  const mockOnChange = jest.fn();
  const mockRenderOption = jest.fn(({ label, ...restProps }) => (
    <div className="search_option" {...restProps}>
      {label}
    </div>
  ));

  it('should render properly', () => {
    const searchInput = render(
      <SearchInput
        loadOptions={mockLoadOptions}
        label="testLabel"
        renderOption={mockRenderOption}
        onChange={mockOnChange}
        value=""
      />
    );

    expect(searchInput.container.firstChild).toMatchSnapshot();
  });

  it('should return expected results', () => {
    const testValue = 'testValue';

    const searchInput = render(
      <SearchInput
        loadOptions={mockLoadOptions}
        label="testLabel"
        //renderOption={mockRenderOption}
        onChange={mockOnChange}
        value=""
      />
    );

    const inputElement = searchInput.container.getElementsByTagName('input')[0];
    const optionElement = searchInput.container.getElementsByClassName('search_option')[0];

    fireEvent.change(inputElement, { target: { value: testValue } });

    // expect(optionElement).toBeDefined();

    // fireEvent.click(optionElement);

    // expect(mockLoadOptions).toBeCalledWith(testValue);
  });
});
