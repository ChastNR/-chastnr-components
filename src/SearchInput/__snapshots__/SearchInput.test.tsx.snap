// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should render properly 1`] = `
<div
  class="ctrl"
>
  <div
    class="ctrl__cntr"
    role="textbox"
    tabindex="0"
  >
    <label
      class="ctrl__lb"
    >
      testLabel
    </label>
    <input
      autocomplete="off"
      id=":r0:"
      type="text"
      value=""
    />
  </div>
  <div
    class="er__cntr"
    style="height: 0px;"
  >
    <div />
  </div>
</div>
`;
